Создал новую ветку controllers-films-users.

ViktorSalk opened the pull request:
Создал пакет model. Реализовал в нем классы: Film и User. Это классы — модели данных приложения. Создал два класса-контроллера. FilmController для фильмов, и UserController для пользователей. Создал UserNotFoundException, ValidationException, GlobalExceptionHandler для валидации. Создал тесты.

Viktor3642 committed:
Создал новую ветку controllers-films-users.
Создал пакет model. Реализовал в нем классы: Film и User. Это классы — модели данных приложения.
Создал два класса-контроллера. FilmController для фильмов, и UserController для пользователей.
Создал UserNotFoundException, ValidationException, GlobalExceptionHandler для валидации.
Создал тесты.

Viktor3642 committed:
Изменил обработку исключений, чтобы пройти проверку POSTMAN test

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/FilmorateApplication.java:
```diff
@@ -3,10 +3,26 @@
 
+		final Gson gson = new Gson();
+		final Scanner scanner = new Scanner(System.in);
+		System.out.print("Введите JSON => ");
```
Код со строк 17-27 нужно удалить

ViktorSalk replied:
1) в FilmorateApplication удалил лишний код.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,92 @@
+    }
+
+    @GetMapping
+    public Collection<Film> findAll() {
```
Лучше возвращать не `Collection` менее абстрактный интерфейс - `List`, так мы точно будем знать какой тип коллекции ожидать в ответе

ViktorSalk replied:
Использовал более конкретный тип List вместо Collection, сделав API более понятным и защитив его от возможных изменений извне.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    @GetMapping
+    public Collection<User> findAll() {
```
Лучше возвращать не `Collection` менее абстрактный интерфейс - `List`, так мы точно будем знать какой тип коллекции ожидать в ответе

ViktorSalk replied:
Использовал более конкретный тип List вместо Collection, сделав API более понятным и защитив его от возможных изменений извне.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/exceptions/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,27 @@
+public class GlobalExceptionHandler {
+
+    @ExceptionHandler(ValidationException.class)
+    public ResponseEntity<Map<String, String>> handleValidationException(ValidationException ex) {
```
- Вместо `Map<String, String>` лучше взвращать объект `ErrorResponse`, в котором будет поле `error`, куда нужно будет записывать причину ошибки
- Можно сделать код еще лаконичнее, избавившись от обертки `ResponseEntity`. Для этого нужно просто добавить аннотацию `@ResponseStatus(HttpStatus. BAD_REQUEST)` над методом, а в самом методе возвращать `ErrorResponse`.

ViktorSalk replied:
Вместо Map<String, String> использовал @ResponseStatus(HttpStatus. BAD_REQUEST), создав ErrorResponse.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,12 +1,15 @@
+@Data
 public class Film {
+
+    Long id;
```
У всех полей класса должен быть модификатор доступа

ViktorSalk replied:
Указал модификатор доступа для полей класса, для геттеров и сеттеров при аннотации использовал приватный.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -0,0 +1,15 @@
+
+@Data
+public class User {
+
```
У всех полей класса должен быть модификатор доступа

ViktorSalk replied:
Указал модификатор доступа для полей класса, для геттеров и сеттеров при аннотации использовал приватный.

afomkina reviewed (changes requested):
Очень рекоммендую попробовать сделать дополнительное задание, так как сприговая валидация очень популярна в коммерческой разработке и никто не пишет валидацию руками. Она очень сильно сокращает код и избавляет от ошибок "невнимательности". Вот [отлчиная статься](https://reflectoring.io/bean-validation-with-spring-boot/#a-custom-validator-with-spring-boot:~:text=Validation%20with%20Spring%20Boot%20%2D%20the%20Complete%20Guide) про аннотации валидации

Viktor3642 committed:
Исправил замечания 10 спринта проекта Filmorate ветки controllers-films-users
1) в FilmorateApplication удалил лишний код.
2) в FilmController и UserController использовал более конкретный тип List вместо Collection сделав API более понятным и защищая от возможных изменений извне.
3) в GlobalExceptionHandler вместо Map<String, String> использовал @ResponseStatus(HttpStatus. BAD_REQUEST) создав ErrorResponse.
4) в Film и User указал модификатор доступа для полей класса, для геттеров и сеттеров при аннотации, использовал приватный.
5) Реализовал дополнительное задание используя спринговую валидацию, С помощью аннотаций задав ограничения, которые будут проверяться автоматически: подредактировал Film и User, FilmController и UserController, FilmControllerTests и UserControllerTests. Чтобы Spring не только преобразовал тело запроса в соответствующий класс, но и проверил корректность переданных данных, вместе с аннотацией @RequestBody использовал аннотацию @Valid.

Viktor3642 committed:
Убрал неиспользуемые импорты в FilmorateApplication для прохождения контроля стиля.

Viktor3642 committed:
Убрал неиспользуемые импорты в UserController для прохождения контроля стиля.

Viktor3642 committed:
Обновил FilmControllerTests и FilmController для прохождения POSTMAN tests.

ViktorSalk commented:
> Очень рекомендую попробовать выполнить дополнительное задание, так как сприговая валидация очень популярна в коммерческой разработке, и никто не пишет валидацию вручную. Она значительно сокращает код и избавляет от «невнимательных» ошибок. Вот [отличная статья](https://reflectoring.io/bean-validation-with-spring-boot/#a-custom-validator-with-spring-boot:~:text=Validation%20with%20Spring%20Boot%20%2D%20the%20Complete%20Guide) об аннотациях валидации

Спасибо) Реализовал дополнительное задание, используя проверку Spring, с помощью аннотаций задав ограничения, которые будут проверяться автоматически: подредактировал классы Film и User, контроллеры FilmController и UserController, тесты FilmControllerTests и UserControllerTests. Чтобы Spring не только преобразовывал тело запроса в соответствующий класс, но и проверял корректность переданных данных, вместе с аннотацией @RequestBody использовал аннотацию @Valid.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,101 @@
+    }
+
+    @PostMapping
+    public ResponseEntity<Film> createFilm(@Valid @RequestBody Film film) {
```
Здесь лучше тоже использовать `@ResponseStatus(HttpStatus.CREATED)`, чтобы избавиться от дополнительной обертки `ResponseEntity` и просто возвращать `Film`

ViktorSalk replied:
В FilmController в createFilm использовал @ResponseStatus(HttpStatus.CREATED), чтобы избавиться от дополнительной обертки.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,101 @@
+    }
+
+    @PutMapping
+    public ResponseEntity<Film> updateFilm(@Valid @RequestBody Film film) {
```
Нужно использовать `@ResponseStatus`

ViktorSalk replied:
В FilmController в updateFilm использовал @ResponseStatus(HttpStatus.ОК).

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/exceptions/ErrorResponse.java:
```diff
@@ -0,0 +1,12 @@
+public class ErrorResponse {
+    private final String error;
+
+    public ErrorResponse(String error) {
```
[@Data](https://projectlombok.org/features/Data) включает в себя аннотацию [@RequiredArgsConstructor](https://projectlombok.org/features/constructor), поэтому явно его объявлять не нужно

ViktorSalk replied:
В ErrorResponse убрал явное объявление.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/exceptions/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,22 @@
+
+    @ExceptionHandler(UserNotFoundException.class)
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    public ErrorResponse handleUserNotFoundException(UserNotFoundException ex) {
```
Лучше явно прописать название переменной, как `exception`. Без сокращений код выглядит более читаемо

ViktorSalk replied:
В GlobalExceptionHandler убрал сокращения.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,12 +1,28 @@
+    private Long id;
+
+    @NotNull(message = "Название фильма не может быть пустым")
+    @NotBlank(message = "Название фильма не может быть пустым")
```
На самом деле [@NotBlank](https://www.baeldung.com/java-bean-validation-not-null-empty-blank#:~:text=greater%20than%20zero.-,%40NotBlank,-%3A%20a%20constrained) содержет проверку на null, поэтому добавлять дополнительно `@NotNull` не нужно

ViktorSalk replied:
Спасибо) В Film убрал лишние @NotNull.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,12 +1,28 @@
+    private String name;
+
+    @NotNull(message = "Описание фильма не может быть пустым")
+    @NotBlank(message = "Описание фильма не может быть пустым")
```
Нужно убрать `@NotNull`

ViktorSalk replied:
В Film убрал лишние @NotNull.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -0,0 +1,28 @@
+
+    private Long id;
+
+    @NotNull(message = "Email не может быть пустым")
```
Нужно удалить лишние `@NotNull` в данном классе

ViktorSalk replied:
В User убрал лишние @NotNull.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,12 +1,28 @@
+
+    @NotNull(message = "Описание фильма не может быть пустым")
+    @NotBlank(message = "Описание фильма не может быть пустым")
+    private String description;
```
Чтобы полностью избавиться от ручной валидации, нужно добавить [@Size](https://www.baeldung.com/java-validation#:~:text=value%20is%20true.-,%40Size,-validates%20that%20the)

ViktorSalk replied:
В Film description добавил @Size на 200 символов.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,12 +1,28 @@
+    private String description;
+
+    @NotNull(message = "Дата выхода не может быть пустой")
+    private LocalDate releaseDate;
```
Можно реализовать кастомную аннотацию валидации, чтобы полностью отказаться от ручной валидации.
Вот [отличная статья](https://reflectoring.io/bean-validation-with-spring-boot/#a-custom-validator-with-spring-boot:~:text=to%20none.-,A%20Custom%20Validator%20with%20Spring%20Boot,-If%20the%20available), в которой подробно описано, как это можно сделать

ViktorSalk replied:
Спасибо) Реализовал кастомную аннотацию валидации ReleaseDateValidation для Film.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -1,12 +1,28 @@
+    private LocalDate releaseDate;
+
+    @NotNull(message = "Длительность не может быть пустой")
+    private Integer duration;
```
[@PositiveOrZero](https://www.baeldung.com/java-validation#:~:text=%40Positive%20and-,%40PositiveOrZero,-apply%20to%20numeric) проверяет, что число больше либо равно нулю

ViktorSalk replied:
В Film duration добавил @PositiveOrZero на длительность фильма.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,80 @@
+        user.setId(getNextId());
+
+        // Проверка на существующий логин
+        if (users.values().stream().anyMatch(existingUser -> existingUser.getLogin().equals(user.getLogin()))) {
```
_Дополнительные оптимизации_
Чтобы проверка почты проходила за [O(1)](https://skillbox.ru/media/code/big-o-notation-chto-eto-takoe-i-kak-eye-poschitat/), нужно завести `Set`, куда при добавлении нового пользователя мы будем класть его `email`. Соответственно, перед тем как добавить пользователя, нужно будет посмотреть, есть ли уже в сете подобный `email`, если есть - то пользователь с таким `email` уже существует.

Так как поиск и вставка в сете работают в среднем за O(1), то наш алгоритм поиска одинаковых `email` тоже будет работать в среднем за O(1), а это гораздо быстрее, чем пройтись по всем элементам списка O(n)

ViktorSalk replied:
В UserController провел дополнительную оптимизацию алгоритма поиска, чтобы проверка почты проходила быстрее, за O(1).

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,80 @@
+            throw new ValidationException("Ошибка. Попытка добавить логин, который уже существует");
+        }
+
+        if (user.getName() == null || user.getName().isBlank()) {
```
Эту проверку можно перенести в конструктор класса `User`, тогда поле будет автоматически проставляться сразу при создании экземпляра класса

ViktorSalk replied:
Перенес проверку из UserController в конструктор класса User, создав метод setLogin.

Viktor3642 committed:
1) В FilmController в createFilm использовал @ResponseStatus(HttpStatus.CREATED), чтобы избавиться от дополнительной обертки.
2) В FilmController в updateFilm использовал @ResponseStatus(HttpStatus.ОК).
3) В ErrorResponse убрал явное объявление.
4) В GlobalExceptionHandler убрал сокращения.
5) В Film убрал лишние @NotNull.
6) В Film description добавил @Size на 200 символов.
7) Реализовал кастомную аннотацию валидации ReleaseDateValidation для Film.
8) В Film duration добавил @PositiveOrZero на длительность фильма.
9) В UserController провел дополнительную оптимизацию алгоритма поиска, чтобы проверка почты проходила быстрее, за O(1).
10) Перенес проверку init из UserController в конструктор класса User.
11) В User убрал лишние @NotNull.

Viktor3642 committed:
1) В классе User изменил аннотацию @PostConstruct, на метод setLogin.

afomkina commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,99 @@
+        throw new UserNotFoundException("Фильм с id = " + film.getId() + " не найден");
+    }
+
+    private void filmValidator(@RequestBody Film film) {
```
Этот метод нужно удалить, так как валидация теперь работает с помощью аннотаций

ViktorSalk replied:
Удалил filmValidator, дублирующий валидацию.

Viktor3642 committed:
Удалил filmValidator, дублирующий валидацию.
